### PR TITLE
[java] New rule: AvoidAssert

### PR DESCRIPTION
**Pull request by nathanspectacular**
## Describe the PR

Adds a rule called AvoidAssert.

For test code, use the JUnit or TestNG Assert class since it cannot be disabled like the assert keyword.

For production code, it requires 3 test cases to cover all scenarios:

1. The assert defaults to disabled.  In this case, the assert's condition is not executed.  
2. If asserts are enabled, the assert's condition can evaluate to true and the condition may introduce side-effects that must be tested.
3. If asserts are enabled, the assert's condition can evaluate to false and the exception handling must be tested.

An if statement is a better choice since it cannot be disabled and hence reduces the need for 1 test case.

## Ready?

- [X] Added unit tests for fixed bug/feature
- [X] Passing all unit tests
- [ ] Complete build `./mvnw clean verify` passes (checked automatically by github actions)
- [X] Added (in-code) documentation (if needed)

Note: This is my first pull request for PMD and Github.  This rule is very simple so that I can learn the process.  Please let me know what else I need to do.  If you feel I need to enhance the rule, please let me know.

### CODE REVIEW

**Comment by oowekyala** [comment]
> The good point about assert statements is, that they indeed *can* be disabled. But then again, they **must not have any side-effect**.

This is exactly my sentiment. I think we should have a rule "SideEffectInAssert" to flag side effects, but pure asserts are absolutely fine and we shouldn't track them down with a generic AvoidAssert.

My suggestion would be to have a rule SideEffectInAssert, and another rule to flag asserts in tests (maybe named "NonTestAssertion") instead of AvoidAssert.

> For test code, use the JUnit or TestNG Assert class since it cannot be disabled like the assert keyword.

I think this justification is pretty thin, since I would say every test runner out there uses `-ea` (or should do so). One more tangible benefit to using those classes is they provide better error messages.

> The first property I propose is annotations. annotations will have JUnit's and TestNG's @Test annotations written as fully qualified names. If any method in the class has these annotations, then the rule will flag all asserts in the class.

I think there's no need to have a property for this. We could just use a list of annotations that belong to those test frameworks, in line with some other rules about testing.

> The second property I propose is paths. paths is a : delimited set of path fragments (e.g. /test/). Any class that has one of these path fragments in the file path will not be allowed to have asserts in the class. The default for this property will be /test/. The end user can add additional paths as desired. For example, I could put /src/ in my projects and disallow asserts in all code including production code.

I think this can be achieved by configuring rulesets. I don't like the idea of adding ad-hoc file filters to rules

**Comment by adangel** [comment]
> I did not check off "Complete build ./mvnw clean verify passes" since it appears to be automated. However, I did do this. In the future, shall I check this off?

This is a checklist for you and us: If you tick this, you say "it worked on my machine, I tested it...".

Since you are a first-time contributor, you pull request must first be approved by one of us, before it is executed on github actions, see https://docs.github.com/en/actions/managing-workflow-runs/approving-workflow-runs-from-public-forks for background info.

**Comment by adangel** [comment]
>> introduces 1884 new violations,

> That is a lot of new violations! Is there something I need to do with this? Is this expected?

Since this is a new rule, new violations are expected.
But the report gives you the opportunity to look through the found violations and think about, whether they are real violations or false positives.

**Comment by adangel** [comment]
>> For test code, use the JUnit or TestNG Assert class since it cannot be disabled like the assert keyword.
> I think this justification is pretty thin, since I would say every test runner out there uses -ea (or should do so).

There is indeed a test runner, that doesn't enable assertions by default (for whatever reason...): https://www.eclipse.org/tycho/sitedocs/tycho-surefire-plugin/test-mojo.html#enableAssertions

The usual maven surefire plugin however, does this by default: https://maven.apache.org/surefire/maven-surefire-plugin/test-mojo.html#enableAssertions

> One more tangible benefit to using those classes is they provide better error messages.

Yes, that's the reason, why you should use a more sophisticated Assert class to report a failing test.

In summary, I think, this rule should be named something like "AssertStatementInTests" and the rule should only report assert statements in test classes. The category here would be best practices.
An example for this would be [GenericConversionServiceTests.java#L422](https://github.com/spring-projects/spring-framework/blob/v5.0.6.RELEASE/spring-core/src/test/java/org/springframework/core/convert/support/GenericConversionServiceTests.java#L422)

And maybe, we should have a separate rule, that finds assert statement in both production code and test code, that have side effects (e.g. "AssertStatementWithSideeffect"). This would belong to the category error prone.
